### PR TITLE
feat: New Module `avm/res/operational-insights/workspace`

### PR DESCRIPTION
**Pull request by krbar**
## Description

New Module `avm/res/operational-insights/workspace`, migrated from CARML.

## Adding a new module

- [x] A proposal has been submitted and approved.
- [ ] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [x] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

| Pipeline |
| - |
| [avm.res.operational-insights.workspace](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.operational-insights.workspace.yml) |

### CODE REVIEW

File: avm/res/operational-insights/workspace/data-export/main.bicep [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+param enableDefaultTelemetry bool = true
+
+@description('Optional. An array of tables to export, for example: [\'Heartbeat\', \'SecurityEvent\'].')
+param tableNames array = []
+
+// =============== //
+//   Deployments   //
+// =============== //
+
+resource defaultTelemetry 'Microsoft.Resources/deployments@2021-04-01' = if (enableDefaultTelemetry) {
```
Comment (AlexanderSehr) [inline avm/res/operational-insights/workspace/data-export/main.bicep:33]: Remove telemetry from all children

Comment (krbar) [reply avm/res/operational-insights/workspace/data-export/main.bicep]: Removed and re-tested, commit https://github.com/Azure/bicep-registry-modules/pull/612/commits/c2dfbf20b99f619db38c62595a68cc8f31cd9812

File: avm/res/operational-insights/workspace/tests/e2e/adv/dependencies.bicep [hunk 2]
```diff
@@ -0,0 +1,85 @@
+@description('Optional. The location to deploy to.')
+param location string = resourceGroup().location
+
+@description('Required. The name of the Storage Account to create.')
+param storageAccountName string
+
+@description('Required. The name of the Automation Account to create.')
+param automationAccountName string
+
+@description('Required. The name of the Event Hub Workspace to create.')
```
Comment (AlexanderSehr) [inline avm/res/operational-insights/workspace/tests/e2e/adv/dependencies.bicep:10]:
```suggestion
@description('Required. The name of the Event Hub Namespace to create.')
```

File: avm/res/operational-insights/workspace/tests/e2e/adv/main.test.bicep [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+}
+
+module nestedDependencies 'dependencies.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-nestedDependencies'
+  params: {
+    location: location
+    storageAccountName: 'dep${namePrefix}sa${serviceShort}'
+    automationAccountName: 'dep-${namePrefix}-auto-${serviceShort}'
+    eventHubNamespaceName: 'dep-${namePrefix}-ehw-${serviceShort}'
```
Comment (AlexanderSehr) [inline avm/res/operational-insights/workspace/tests/e2e/adv/main.test.bicep:40]:
```suggestion
    eventHubNamespaceName: 'dep-${namePrefix}-ehn-${serviceShort}'
```

File: avm/res/operational-insights/workspace/tests/e2e/waf-aligned/main.test.bicep [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    eventHubNamespaceName: 'dep-${namePrefix}-evhns-${serviceShort}'
+    location: location
+  }
+}
+
+// ============== //
+// Test Execution //
+// ============== //
+
+module testDeployment '../../../main.bicep' = {
```
Comment (AlexanderSehr) [inline avm/res/operational-insights/workspace/tests/e2e/waf-aligned/main.test.bicep:63]: I assume this is a copy of the common test. Any chance we could throw out everything that isn't required to be WAF / PSRule aligned? E.g. Role assignments etc. to only surface the content that is relevant to understand the explicit example

File: avm/res/operational-insights/workspace/tests/e2e/waf-aligned/dependencies.bicep [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+@description('The resource ID of the created Storage Account.')
+output storageAccountResourceId string = storageAccount.id
+
+@description('The resource ID of the created Automation Account.')
+output automationAccountResourceId string = automationAccount.id
+
+@description('The principal ID of the created Managed Identity.')
+output managedIdentityPrincipalId string = managedIdentity.properties.principalId
```
Comment (AlexanderSehr) [inline avm/res/operational-insights/workspace/tests/e2e/waf-aligned/dependencies.bicep:44]: If the role assignment were to be removed from the test, this output could be removed too

File: avm/res/operational-insights/workspace/main.bicep [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+  Owner: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '8e3af657-a8ff-443c-a75c-2fe8c4bcb635')
+  Reader: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
+  'Role Based Access Control Administrator (Preview)': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'f58310d9-a9f6-439a-9e8d-f62e7b41a168')
+  'Security Admin': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'fb1c8493-542b-48eb-b624-b4c8fea62acd')
+  'Security Reader': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '39bc4728-0917-49c7-9d2c-d95423bc2eb4')
+  'User Access Administrator': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '18d7d88d-d35e-4fb5-a5c3-7773c20a72d9')
+}
+
+resource avmTelemetry 'Microsoft.Resources/deployments@2023-07-01' = if (enableTelemetry) {
+  name: '46d3xbcp.res.operationsmanagement-solution.${replace('-..--..-', '.', '-')}.${substring(uniqueString(deployment().name, location), 0, 4)}'
```
Comment (AlexanderSehr) [inline avm/res/operational-insights/workspace/main.bicep:122]: Should point to workspace, not solution

File: avm/res/operational-insights/workspace/main.bicep [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+    plan: contains(table, 'plan') ? table.plan : 'Analytics'
+    schema: contains(table, 'schema') ? table.schema : {}
+    retentionInDays: contains(table, 'retentionInDays') ? table.retentionInDays : -1
+    totalRetentionInDays: contains(table, 'totalRetentionInDays') ? table.totalRetentionInDays : -1
+    restoredLogs: contains(table, 'restoredLogs') ? table.restoredLogs : {}
+    searchResults: contains(table, 'searchResults') ? table.searchResults : {}
+  }
+}]
+
+module logAnalyticsWorkspace_solutions '../../operations-management/solution/main.bicep' = [for (gallerySolution, index) in gallerySolutions: if (!empty(gallerySolutions)) {
```
Comment (AlexanderSehr) [inline avm/res/operational-insights/workspace/main.bicep:281]: Should reference the published module and requires a regeneration of the files

File: avm/res/operational-insights/workspace/main.bicep (hunk 6 above)
Comment (krbar) [reply avm/res/operational-insights/workspace/main.bicep]: Absolutely correct, fixed.

File: avm/res/operational-insights/workspace/main.bicep (hunk 7 above)
Comment (krbar) [reply avm/res/operational-insights/workspace/main.bicep]: Done

File: avm/res/operational-insights/workspace/tests/e2e/waf-aligned/dependencies.bicep (hunk 5 above)
Comment (krbar) [reply avm/res/operational-insights/workspace/tests/e2e/waf-aligned/dependencies.bicep]: Removed MSI completely from the dependencies, as it is not in use any more: https://github.com/Azure/bicep-registry-modules/pull/612/commits/e847310eec90c9e7c72e37777984fe6939e01398

File: avm/res/operational-insights/workspace/tests/e2e/waf-aligned/main.test.bicep (hunk 4 above)
Comment (krbar) [reply avm/res/operational-insights/workspace/tests/e2e/waf-aligned/main.test.bicep]: I removed some stuff from this test, please check if it's fine

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**